File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current =  head
        previous = None
        next_node = None

        while current is not None:
            next_node = current.next ## remeber next on the list
            current.next = previous ## reverse arrow direction
            previous = current ### update our previous to next ListNode
            current = next_node    ## update our current cycle

        return previous

        

File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head -> next;
        while (fast!= nullptr && fast -> next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // splitting the node into 2 halves
        ListNode* second = slow->next;
        slow -> next = nullptr;

        // we reverse our second half
        ListNode* curr = second;
        ListNode* prev = nullptr;

        while (curr!= nullptr) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }

        second = prev;

        ListNode* first = head;

        while (second!= nullptr) {
        ListNode* temp1 = first->next;
        ListNode* temp2 = second->next;

        first->next = second;
        second->next = temp1;

        first = temp1;
        second = temp2;
            }
    }
};
